### PR TITLE
Fixed #24605 -- incorrect reference to alias in subquery

### PR DESCRIPTION
**Pull request by akaariai**
Thanks to charettes and priidukull for investigating the issue, and to
kurevin for the report.

### CODE REVIEW

File: django/db/models/sql/compiler.py [hunk 1]
```diff
@@ -340,7 +340,8 @@ def quote_name_unless_alias(self, name):
         if name in self.quote_cache:
             return self.quote_cache[name]
         if ((name in self.query.alias_map and name not in self.query.table_map) or
-                name in self.query.extra_select or name in self.query.external_aliases):
+                name in self.query.extra_select or (
+                    name in self.query.external_aliases and name not in self.query.table_map)):
```
Comment (charettes) [inline django/db/models/sql/compiler.py]: Ahh we were so close, we tried removing `or name in self.query.external_aliases` but this triggered more failures. This all makes sense now.

**Comment by charettes** [comment]
This actually solves the reference error but it looks like filter ordering still produce different subquery contraints.

`WHERE U1."id" IS NULL` compared to `WHERE (U1."id" IS NULL AND U0."id" = ("Individual"."id"))`

Is this issue already tracked somewhere else? The fact that some queries can differ based on filter ordering looks like a definitive footgun to me.

**Comment by akaariai** [comment]
That one is an artifact of hacks trying to support multiple m2m filters in one filter clause. I think https://github.com/django/django/pull/4385 will solve this one properly. Until that is committed it is better to leave this part of code alone. The results aren't different in this case, even if the queries are different.

**Comment by charettes** [comment]
Alright, I'll mark this patch as RFC in this case. Do you think it could be worth adding an additional test case in #4385 to make sure it is not filter ordering dependent?

**Comment by timgraham** [comment]
Release notes added and merged in 355c5edd9390caad5725375abca03460805f663b, thanks!

**Pull request closed without merging**